File: others/프로그래머스_베스트앨범.py
def solution(genres, plays):

    answer = []

    play_cnt = {}
    genre_sort = {}

    for (idx, genre) in enumerate(genres):
        if genre not in play_cnt:
            play_cnt[genre] = 0
            genre_sort[genre] = []

        genre_sort[genre].append([plays[idx], idx])
        play_cnt[genre] += plays[idx]

    sorted_cnt = sorted(play_cnt.items(), key=lambda x: x[1], reverse=True)

    for each in sorted_cnt:
        genre = each[0]
        list = sorted(genre_sort[genre], key=lambda x: x[1], reverse=True)
        list.sort(key=lambda x: x[0])
        n = 0
        while list:
            to_add = list.pop()
            answer.append(to_add[1])
            n += 1

            if n == 2:
                break

    return answer

File: others/프로그래머스_베스트앨범.js
function solution(genres, plays) {
  const answer = [];

  // 장르 별 재생 횟수 더하기
  const playCnt = Object()
  // 각 장르마다 재생횟수/ idx 리스트로 저장
  const genreSort = Object()

  genres.forEach(function(genre, idx) {
    if  (!(genre in playCnt)) {
      playCnt[genre] = 0
      genreSort[genre] = Array()
    }

    playCnt[genre] += plays[idx]
    genreSort[genre].push([plays[idx], idx])
  })

  // 재생 많이 된 순으로 정렬
  const sortedCnt = Object.entries(playCnt).sort(function (a, b) { return b[1]-a[1]})

  // 재생 많이 된 장르 순으로 2개씩 앨범에 추가
  sortedCnt.forEach(genre => {
    const toAdd = genreSort[genre[0]].sort(function(a, b) {return b[0]-a[0]})

    for (let i=0; i < 2; i++) {
      if (toAdd.length === i) {
        break
      }
      else {
        answer.push(toAdd[i][1])
      }
    }
  })

  return answer;
}

console.log(solution(["classic", "pop", "classic", "classic", "pop"], [500, 600, 150, 800, 2500]))